src/ssh/sftp_client.cpp

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (C) 2019 Canonical, Ltd.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 3.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <multipass/ssh/sftp_client.h>
+#include <multipass/ssh/throw_on_error.h>
+#include <multipass/utils.h>
+
+#include "ssh_client_key_provider.h"
+
+#include <fmt/format.h>
+
+#include <array>
+#include <fcntl.h>
+
+#include <QFile>
+
+namespace mp = multipass;
+
+namespace
+{
+constexpr int file_mode = 0664;
+constexpr auto max_transfer = 65536u;
+const std::string stream_file_name{"stream_output.dat"};
+
+using SFTPFileUPtr = std::unique_ptr<sftp_file_struct, int (*)(sftp_file)>;
+using SFTPAttrUPtr = std::unique_ptr<sftp_attributes_struct, void (*)(sftp_attributes)>;
+
+mp::SFTPSessionUPtr make_sftp_session(ssh_session session)
+{
+    mp::SFTPSessionUPtr sftp{sftp_new(session), sftp_free};
+
+    if (sftp == nullptr)
+        throw std::runtime_error(fmt::format("could not create new sftp session: {}", ssh_get_error(session)));
+
+    return sftp;
+}
+
+std::string full_destination(const std::string& destination_path, const std::string& filename)
+{
+    if (destination_path.empty())
+    {
+        return filename;
+    }
+    else if (mp::utils::is_dir(destination_path))
+    {
+        return fmt::format("{}/{}", destination_path, filename);
+    }
+
+    return destination_path;
+}
+} // namespace
+
+mp::SFTPClient::SFTPClient(const std::string& host, int port, const std::string& username,
+                           const std::string& priv_key_blob)
+    : SFTPClient{std::make_unique<mp::SSHSession>(host, port, username, mp::SSHClientKeyProvider(priv_key_blob))}
+{
+}
+
+mp::SFTPClient::SFTPClient(SSHSessionUPtr ssh_session)
+    : ssh_session{std::move(ssh_session)}, sftp{make_sftp_session(*this->ssh_session)}
+{
+    SSH::throw_on_error(sftp, *this->ssh_session, "[sftp pull] init failed", sftp_init);
+}
+
+void mp::SFTPClient::push_file(const std::string& source_path, const std::string& destination_path)
+{
+    auto full_destination_path = full_destination(destination_path, mp::utils::filename_for(source_path));
+    SFTPFileUPtr file_handle{
+        sftp_open(sftp.get(), full_destination_path.c_str(), O_WRONLY | O_CREAT | O_TRUNC, file_mode), sftp_close};
```

Comment:
My only concern is how easily we overwrite files. Or am I being too nanny-ish?

Reply:
So did an experiment to see what the actual `sftp` client does and it will just overwrite the old file with the new file without asking.  So I think it's ok to leave as-is, but perhaps if it becomes an issue, we can do something about it in the future.